dispatcher.go

```diff
@@ -103,16 +111,34 @@ func NewDispatcher(cfg Config) *Dispatcher {
 		panic("yarpc.NewDispatcher expects a valid service name: %s" + err.Error())
 	}
 
+	logger := zap.NewNop()
+	if cfg.logger != nil {
+		logger = cfg.logger.Named("yarpc").With(
+			zap.Namespace("yarpc"), // isolate yarpc's keys
+			zap.String("dispatcher", cfg.Name),
+		)
+		cfg = addObservingMiddleware(cfg, logger)
+	}
+
 	return &Dispatcher{
 		name:              cfg.Name,
 		table:             middleware.ApplyRouteTable(NewMapRouter(cfg.Name), cfg.RouterMiddleware),
 		inbounds:          cfg.Inbounds,
 		outbounds:         convertOutbounds(cfg.Outbounds, cfg.OutboundMiddleware),
 		transports:        collectTransports(cfg.Inbounds, cfg.Outbounds),
 		inboundMiddleware: cfg.InboundMiddleware,
+		log:               logger,
 	}
 }
 
+func addObservingMiddleware(cfg Config, logger *zap.Logger) Config {
+	cfg.InboundMiddleware.Unary = inboundmiddleware.InboundChain(observerware.NewUnaryInbound(
+		logger,
+		nil, // observerware.ContextExtractor
```

Comment:
We need to pass a NopContextExtractor here, or setup a default in the NewUnaryInbound function, otherwise I think this will panic when we try to extract Context info.